core/silkworm/consensus/clique/clique_snapshot.cpp

```diff
@@ -0,0 +1,249 @@
+   limitations under the License.
+*/
+
+#include "clique_snapshot.hpp"
+#include <silkworm/common/endian.hpp>
+#include <silkworm/common/util.hpp>
+#include <silkworm/crypto/ecdsa.hpp>
+#include <cstring>
+
+namespace silkworm {
+
+std::array<uint8_t, 8> kNonceAuthorize =   {0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff};
+std::array<uint8_t, 8> kNonceUnauthorize = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+
+//! \brief Updated snapshot by adding headers
+//! \param headers: list of headers to add.
+ValidationResult CliqueSnapshot::add_header(const BlockHeader& header, const evmc::address& signer, const CliqueConfig& config) {
+    auto hash{header.hash()};
+    evmc::address allowed_signer{}; // We only modify snapshot after checks are done.
+    // Delete the oldest signer from the recent list to allow it signing again
+    if (recents_.size() > signers_.size() / 2) {
+        allowed_signer = recents_.back();
+    }
+
+    if (std::find(signers_.begin(), signers_.end(), signer) == signers_.end()) {
+        return ValidationResult::kUnauthorizedSigner;
+    }
+
+    // Remove any votes on checkpoint blocks
+    if (header.number % config.epoch == 0) {
+        tallies_.clear();
+    }
+
+    if (std::find(recents_.begin(), recents_.end(), signer) != recents_.end() && signer != allowed_signer) {
+        return ValidationResult::kRecentlySigned;
+    }
+
+    // We check what the vote is
+    bool authorize;
+    if (header.nonce == kNonceAuthorize) {
+        authorize = true;
+    } else if (header.nonce == kNonceUnauthorize) {
+        authorize = false;
+    } else {
+        return ValidationResult::kInvalidVote;
+    }
+
+    if (allowed_signer) {
+        recents_.pop_back();
+    }
+    recents_.push_front(signer);
+
+    if (!header.beneficiary) {
+        update(header.number, hash);
+        return ValidationResult::kOk;
+    }
+
+    // Uncast votes from signers
+    decrement_vote(header.beneficiary, signer);
+    // do casting
+    increment_vote(header.beneficiary, signer, authorize);
+    // If the vote passed, update the list of signers
+    auto current_tally{tallies_[header.beneficiary]};
+    if (current_tally.votes > signers_.size() / 2) {
+        if (current_tally.authorize) {
+            signers_.push_back(header.beneficiary);
+        } else {
+            signers_.erase(std::remove(signers_.begin(), signers_.end(), header.beneficiary), signers_.end());
+            // Clean up recents
+            recents_.pop_back();
+            // Update Tallies
+            clear_votes(header.beneficiary);
+        }
+        if (tallies_.count(header.beneficiary)) {
+            tallies_.erase(header.beneficiary);
+        }
+    }
+
+    update(header.number, hash);
+    // Success
+    return ValidationResult::kOk;
+}
+
+//! \brief Verify seal for header
+//! \param header: header to verify.
+ValidationResult CliqueSnapshot::verify_seal(const BlockHeader& header, const evmc::address& signer) {
```

Comment:
This whole function is not covered by tests which means test cases never verify seal.